Add live migration support on MSHV

russell-islam opened the pull request:
This PR adds the APIs needed for live migrations on MSHV. Modifies existing get_dirty_log function to accept extra parameters needed for MSHV. Implement start/stop_dirty_log for MSHV. For this PR I have tested Live migrations and measures some numbers. The numbers are very close to KVM.  Steps used to test this on MSHV:

Source VM:

sudo ./cloud-hypervisor/target/release/cloud-hypervisor --kernel ./linux-cloud-hypervisor/arch/x86/boot/compressed/vmlinux.bin --disk path=focal-server-cloudimg-amd64.raw --cpus boot=4 --memory size=1G --cmdline "root=/dev/vda1 console=ttyS0" --serial tty --console off --api-socket=/tmp/api1 


Destination VM: 
sudo ./cloud-hypervisor/target/release/cloud-hypervisor --api-socket=/tmp/api2

#Receiving migration for the destination

sudo ./cloud-hypervisor/target/release/ch-remote --api-socket=/tmp/api2 receive-migration unix:/tmp/sock

#Send migration
time sudo ./cloud-hypervisor/target/release/ch-remote --api-socket=/tmp/api1 send-migration unix:/tmp/sock

likebreath commented on vmm/src/vm.rs:
```diff
@@ -2152,6 +2152,25 @@ impl Vm {
     }
 
+    #[cfg(all(feature = "mshv", target_arch = "x86_64"))]
+    pub fn enable_memory_log_dirty(&self) -> std::result::Result<(), MigratableError> {
```
This function can be merged to `Vm::start_memory_dirty_log()` by calling `hypervisor::Vm::enable_dirty_page_tracking()` directly with feature gate.

likebreath commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -884,6 +884,22 @@ impl vm::Vm for MshvVm {
+    ///
+    fn enable_dirty_page_tracking(&self) -> vm::Result<()> {
+        self.fd
+            .enable_dirty_page_tracking()
```
Will this enable dirty-page-tracking on all guest memory regions? 

With KVM, we are now enabling dirty-page-tracking only on memory regions for guest RAM and excluding regions backed by device/file, e.g. virtio-fs, pmem, SGX. Will mshv be able to do something similar?

/cc @rbradford @sboeuf

russell-islam replied:
Tracking should work for any page mapped to the guest's GPA space. That means whatever memory we map through set_user_memory should be tracked.

russell-islam replied:
How does the map of these types of memory work in CLH?

likebreath replied:
There is no need to track all pages mapped to the guest's GPA space, and we only need to kvm/mshv to track dirty pages for guest RAM. This is how CLH is working with kvm by using mappings from `MemoryManager::guest_ram_mappings`.

Can you make the `mshv` crate provide a API so that it can enable dirty page logging on a slot/region basis (similar to what `kvm_ioctls::ioctls::vm::VmFd::get_dirty_log()` provides)?

russell-islam replied:
Right now Microsoft Hypervisor  does not have such API.

likebreath commented on vmm/src/memory_manager.rs:
```diff
@@ -1504,7 +1507,7 @@ impl MemoryManager {
             let vm_dirty_bitmap = self
                 .vm
-                .get_dirty_log(r.slot, r.gpa, r.size, 0)
+                .get_dirty_log(r.slot, r.gpa, r.size, DIRTY_BITMAP_CLEAR_DIRTY)
```
I am assuming `mshv` will get and reset the dirty-page bitmap with `DIRTY_BITMAP_CLEAR_DIRTY`, right? 

It ensures that there is no racing and no potential missing dirty-page-tracking.

russell-islam replied:
Yes. It ensures that.

likebreath commented on vmm/src/memory_manager.rs:
```diff
@@ -1582,6 +1585,21 @@ impl MemoryManager {
+    pub fn end_memory_dirty_log(&self) -> std::result::Result<(), MigratableError> {
+        for r in &self.guest_ram_mappings {
+            self.vm
+                .get_dirty_log(r.slot, r.gpa, r.size, DIRTY_BITMAP_SET_DIRTY)
```
Can you elaborate more on what 's the behavior of using `DIRTY_BITMAP_SET_DIRTY` (comparing with `DIRTY_BITMAP_CLEAR_DIRTY`)? Also why is it needed?

likebreath replied:
Note that: `guest_ram_mappings` only contains memory regions for guest RAM (a subset of all guest memory regions kvm/mshv created).

Is it intended that only this "subset" of the guest memory regions requires a call to mshv with `DIRTY_BITMAP_SET_DIRTY`? What about the rest memory regions?

russell-islam replied:
Set dirty tells the hypervisor that it either completes the migration or aborts so that hypervisor can rebuild the large pages afterward. While we enabled the tracking hypervisor breaks the large pages.

likebreath replied:
I see. Then it sounded like you will want to do "set dirty" for all regions, as you enabled the tracking for all regions from `enable_dirty_page_tracking` (as you explained above).

Sounds like "set dirty" should be a part of `disable_dirty_page_tracking`. Any reason not do this as a part of `hypervisor::vm::disable_dirty_page_tracking()`?

russell-islam replied:
Sorry. One correction here. Hypervisor breaks the larges upon clearing the bits, not during the enable_tracking.
So disbale_tracking does not have any info on whether bits were cleared or not. But if you clear the birts you must set the bits before disabling the tracking.

likebreath replied:
Similar to @rbradford's comments, I don't understand why we have to expose those `mshv` flags to the `vmm` crate, and they should be encapsulated at least in the `hypervisor` crate. Like what I asked earlier, why it can't be done as a part of `hypervisor::vm::disable_dirty_page_tracking()`?

russell-islam replied:
disable_dirty_page_tracking does not have any info to call get_dirty_log like gps, size etc.

likebreath replied:
Well. That's why I proposed to add the following two methods to the `hypervisor::Vm` Trait in #2894. It aims to support both `kvm` and `mshv` while hiding hypervisor-specific information (say control flags for `kvm` and `mshv`:

```
    /// Start logging dirty pages
    fn start_dirty_log(
        &self,
        slot: u32,
        guest_phys_addr: u64,
        memory_size: u64,
        userspace_addr: u64,
    ) -> Result<()>;
    /// Stop logging dirty pages
    fn stop_dirty_log(
        &self,
        slot: u32,
        guest_phys_addr: u64,
        memory_size: u64,
        userspace_addr: u64,
    ) -> Result<()>;
```

I can understand you will need to expand the current `hypervisor::Vm::get_dirty_log()` method with taking `gpa` as input for `mshv`. Anything else is missing?

russell-islam replied:
I think this is enough.

likebreath commented on vmm/src/seccomp_filters.rs:
```diff
@@ -141,6 +141,8 @@ mod mshv {
     pub const MSHV_RUN_VP: u64 = 0x8100_b807;
     pub const MSHV_GET_VP_STATE: u64 = 0xc028_b80a;
     pub const MSHV_SET_VP_STATE: u64 = 0xc028_b80b;
+    pub const MSHV_SET_PARTITION_PROPERTY: u64 = 0x4010_b80c;
```
Please squash this commit to the previous one.

likebreath commented on vmm/src/lib.rs:
```diff
@@ -1078,6 +1080,9 @@ impl Vmm {
             info!("Migration complete");
+            // Disable dirty logging
+            #[cfg(all(feature = "mshv", target_arch = "x86_64"))]
+            vm.complete_live_migration()?;
```
I think it is better to rename this function to `vm.complete_memory_dirty_log` and call it before `Request::complete().write_to(&mut socket)?;`.

likebreath reviewed (changes requested):
@russell-islam Thanks for the PR. It looks good in general. Please find my comments/questions below. Any thoughts please don't hesitate to share.

likebreath commented:
Also, please don't forget to do spell check on the commit message and comments. :)

russell-islam commented on hypervisor/src/vm.rs:
```diff
@@ -281,7 +281,13 @@ pub trait Vm: Send + Sync {
     fn set_state(&self, state: VmState) -> Result<()>;
     /// Get dirty pages bitmap
-    fn get_dirty_log(&self, slot: u32, memory_size: u64) -> Result<Vec<u64>>;
+    fn get_dirty_log(
+        &self,
+        slot: u32,
+        _base_gpa: u64,
+        memory_size: u64,
+        _flags: u64,
+    ) -> Result<Vec<u64>>;
```
We do have two separate definitions in the VM trait that is confusing too.

likebreath replied:
You can use feature gate around the argument directly, which is commonly used in the current code base. For example: `Vcpu:configure()`: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/48d1a31e087c68aa4c2a7d321e3be06274cc2a73/vmm/src/cpu.rs#L249-L256

russell-islam replied:
Agreed. That way is more readable and clear. Thanks!!

sboeuf commented on hypervisor/src/vm.rs:
```diff
@@ -196,6 +196,16 @@ pub enum HypervisorVmError {
     #[error("Failed to initialize memory region TDX: {0}")]
     InitMemRegionTdx(#[source] std::io::Error),
+    ///
+    /// Error enabling dirt page tracking
```
```suggestion
    /// Error enabling dirty page tracking
```

russell-islam replied:
Fixed

sboeuf commented on hypervisor/src/vm.rs:
```diff
@@ -196,6 +196,16 @@ pub enum HypervisorVmError {
+    #[error("Failed to enable dirty page tracking: {0}")]
+    EnableDirtyPageTracking(#[source] std::io::Error),
+    ///
+    /// Error disabling dirt page tracking
```
```suggestion
    /// Error disabling dirty page tracking
```

russell-islam replied:
Fixed

sboeuf reviewed (commented):
Looks mostly good to me! Please address @likebreath comments and we'll be good to merge this :)

rbradford commented on hypervisor/src/kvm/mod.rs:
```diff
@@ -386,7 +386,13 @@ impl vm::Vm for KvmVm {
     /// Get dirty pages bitmap (one bit per page)
     ///
-    fn get_dirty_log(&self, slot: u32, memory_size: u64) -> vm::Result<Vec<u64>> {
+    fn get_dirty_log(
+        &self,
+        slot: u32,
+        _base_gpa: u64,
+        memory_size: u64,
+        _flags: u64,
+    ) -> vm::Result<Vec<u64>> {
```
Of course that type of compile time behaviour makes it harder to go move to the single mshv + kvm binary in the future. What is the purpose of the flags? Is it MSHV specific? If so, why is that being leaked via the hypervisor API, shouldn't it be encapsulated inside?

russell-islam replied:
Yes. This flag is MSHV specific. We can make this feature guarded. How do we encapsulate it inside while it needs to be provided by the caller?

likebreath replied:
You should not need the caller (from the `vmm` crate) to provide this flag. For example, for the `mshv` implementation of the `hypervisor::Vm` trait, you can do:

1. Always use flag `DIRTY_BITMAP_CLEAR_DIRTY ` from  `hypervisor::Vm::get_dirty_log()`;
2. Use flag `DIRTY_BITMAP_SET_DIRTY` as a part of disabling/stopping dirty pages logging, say from `hypervisor::vm::disable_dirty_page_tracking()`.

Anything I am missing here?

russell-islam replied:
That can be done with your new API.

russell-islam commented:
@likebreath  Please have another look.

likebreath commented:
@russell-islam Thanks for updating the PR. It looks more organized now. I believe the main open here is we don't want to leak hypervisor-specific flags out of the `hypervisor` crate. 

As we discussed above, can you please rebase your PR on #2894 and try to use the updated methods from the `hypervisor::Vm` trait? I assume the two new methods `start_dirty_log()` and `stop_dirty_log()` should be directly reusable by `mshv`, while you will need to expand the `get_dirty_log()` with taking additional argument `gpa` . Please let me know if you encounter problems or have any questions.

russell-islam commented:
@likebreath I have rebased your branch.

rbradford commented:
@liuw Can you confirm that you cannot control the MSHV dirty page logging on a per region basis? That will cause problems for certain regions that are mapped into the guest like `virtio-pmem` or `virtio-fs` cache.

@russell-islam If that truly is the case then please refactor the code in `hypervisor::Vm` to make a single on/off toggle for page tracking and in the KVM case track the regions that need to be enabled/disabled inside the abstraction itself (rather than the `MemoryManager` looping over the RAM regions.) That way we do not have hypervisor specific logic leaking into the VMM core.

liuw commented:
> @liuw Can you confirm that you cannot control the MSHV dirty page logging on a per region basis? That will cause problems for certain regions that are mapped into the guest like `virtio-pmem` or `virtio-fs` cache.

Muminul is correct -- we saw the same specification. MSHV at the moment only has a global control. Can you elaborate a bit what ill effect does that have for `virtio-pmem` and `virtio-fs`?

rbradford commented:
> > @liuw Can you confirm that you cannot control the MSHV dirty page logging on a per region basis? That will cause problems for certain regions that are mapped into the guest like `virtio-pmem` or `virtio-fs` cache.
> 
> Muminul is correct -- we saw the same specification. MSHV at the moment only has a global control. Can you elaborate a bit what ill effect does that have for `virtio-pmem` and `virtio-fs`?

I think that you'll find it will fail. Those memory regions are not included in the `GuestMemoryMmap` which is the method that the RAM is read from/written to in order to accumulate the dirty page data and to fill it in at the destination. They are not included in there as they are not conventional RAM regions and there is no need for the VMM to be able to write to them (e.g. for something like the virtio queues.)

russell-islam commented:
@sboeuf 

> @liuw Can you confirm that you cannot control the MSHV dirty page logging on a per region basis? That will cause problems for certain regions that are mapped into the guest like `virtio-pmem` or `virtio-fs` cache.
> 
> @russell-islam If that truly is the case then please refactor the code in `hypervisor::Vm` to make a single on/off toggle for page tracking and in the KVM case track the regions that need to be enabled/disabled inside the abstraction itself (rather than the `MemoryManager` looping over the RAM regions.) That way we do not have hypervisor specific logic leaking into the VMM core.

@likebreath Any changes on your PR?

likebreath commented:
> >  If that truly is the case then please refactor the code in `hypervisor::Vm` to make a single on/off toggle for page tracking and in the KVM case track the regions that need to be enabled/disabled inside the abstraction itself (rather than the `MemoryManager` looping over the RAM regions.) That way we do not have hypervisor specific logic leaking into the VMM core.
> 
> @likebreath Any changes on your PR?

@russell-islam The PR #2894 of dynamic control over dirty-log has no changes and was just landed this morning. I am working on what @rbradford advised here and will send the PR later today. Please stay tuned.

likebreath commented:
@russell-islam As the PR #2906 is waiting for landing , can you please rebase your PR based on that one and try to use the updated `start/stop_dirty_log` APIs for MSHV (e.g., hiding MSHV specific behaviors inside those APIs)? 

Besides extending the `get_dirty_log` function with an extra argument `gpa` (and seccomp), I assume you don't need to change anything outside of the `hypervisor` crate. Please let me if  there is further questions need to be clarified. Thanks for your efforts!

likebreath commented:
@russell-islam The PR https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2906 is landed.

likebreath commented on hypervisor/src/vm.rs:
```diff
@@ -297,7 +297,12 @@ pub trait Vm: Send + Sync {
+    fn get_dirty_log(
+        &self,
+        slot: u32,
+        #[cfg(feature = "mshv")] base_gpa: u64,
```
As we now only need one new argument `gpa` for MSHV (and it is not MSHV-specific information), I think we can drop the feature gate here (like what we did with the argument `slot`). It makes it easier to support KVM and MSHV in a single binary in the future. WDYT?

likebreath commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -929,6 +932,22 @@ impl vm::Vm for MshvVm {
+    ///
+    /// Enable dirty page tracking
+    ///
+    fn enable_dirty_page_tracking(&self) -> vm::Result<()> {
```
I don't think this API is needed, as  `self.fd.enable_dirty_page_tracking()` is better to be called directly from `start_dirty_log()`.

likebreath commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -929,6 +932,22 @@ impl vm::Vm for MshvVm {
+        self.fd
+            .disable_dirty_page_tracking()
+            .map_err(|e| vm::HypervisorVmError::DisableDirtyPageTracking(e.into()))
+    }
```
Similarly, `self.fd.disable_dirty_page_tracking()` is better to be called directly from `stop_dirty_log()`. In this way, you won't need to change the logic from the `vmm` crate at all.

likebreath commented on vmm/src/vm.rs:
```diff
@@ -2156,12 +2156,40 @@ impl Vm {
     }
 
+    #[cfg(all(feature = "mshv", target_arch = "x86_64"))]
+    pub fn enable_memory_log_dirty(&self) -> std::result::Result<(), MigratableError> {
+        self.vm.enable_dirty_page_tracking().map_err(|e| {
+            MigratableError::EnableDirtyLogging(anyhow!(
+                "Error enabling dirty page tracking: {}",
+                e
+            ))
+        })
+    }
+    #[cfg(all(feature = "mshv", target_arch = "x86_64"))]
+    pub fn disable_memory_log_dirty(&self) -> std::result::Result<(), MigratableError> {
+        self.vm.disable_dirty_page_tracking().map_err(|e| {
+            MigratableError::EnableDirtyLogging(anyhow!(
+                "Error disabling dirty page tracking: {}",
+                e
+            ))
+        })
+    }
+
     pub fn start_memory_dirty_log(&self) -> std::result::Result<(), MigratableError> {
+        #[cfg(all(feature = "mshv", target_arch = "x86_64"))]
+        self.enable_memory_log_dirty()?;
         self.memory_manager.lock().unwrap().start_memory_dirty_log()
     }
 
     pub fn stop_memory_dirty_log(&self) -> std::result::Result<(), MigratableError> {
-        self.memory_manager.lock().unwrap().stop_memory_dirty_log()
+        self.memory_manager
+            .lock()
+            .unwrap()
+            .stop_memory_dirty_log()?;
+
+        #[cfg(all(feature = "mshv", target_arch = "x86_64"))]
+        self.disable_memory_log_dirty()?;
+        Ok(())
     }
```
Changes to the `vmm` crate (e.g.  to "vm.rs") should not be needed, as all MSHV specific operations should be encapsulated inside the `hypervisor` crate with its APIs "start/stop_dirty_log()" (like what the KVM backend does).

russell-islam replied:
Sorry I missed this.

likebreath commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -905,17 +906,19 @@ impl vm::Vm for MshvVm {
+        let dirty_log_slots = self.dirty_log_slots.read().unwrap();
+        for (_, s) in dirty_log_slots.iter() {
+            self.fd
+                .get_dirty_log(s.guest_pfn, s.memory_size as usize, DIRTY_BITMAP_SET_DIRTY)
```
Please add some comments to explain why this is needed.

likebreath commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -807,6 +814,17 @@ impl vm::Vm for MshvVm {
     /// Creates a guest physical memory region.
     fn create_user_memory_region(&self, user_memory_region: MemoryRegion) -> vm::Result<()> {
+        // We only keep the dirty log slots if it is not read only
+        if (user_memory_region.flags & HV_MAP_GPA_WRITABLE) != 0 {
```
Looks to be a typo. Shouldn't it be `if (user_memory_region.flags & HV_MAP_GPA_WRITABLE) != HV_MAP_GPA_WRITABLE` ?

likebreath commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -905,17 +906,27 @@ impl vm::Vm for MshvVm {
-            "functionality not implemented"
-        )))
+        let dirty_log_slots = self.dirty_log_slots.read().unwrap();
+        // Before disabling the dirty page tracking we need
+        // to set the dirty bits in the Hypervisor
+        // This is a requirement from Microsoft Hypervisor
```
Question: what about the "read-only" regions?

russell-islam replied:
No matter read-only or not we keep tracking of the slots. For read-only hypervisor can enable the dirty bits,
        but a VM exit happens before setting the dirty bits. I will change the commit.

likebreath commented on hypervisor/src/vm.rs:
```diff
@@ -206,6 +206,16 @@ pub enum HypervisorVmError {
     ///
     #[error("Failed to initialize memory region TDX: {0}")]
     InitMemRegionTdx(#[source] std::io::Error),
+    ///
+    /// Error enabling dirty page tracking
+    ///
+    #[error("Failed to enable dirty page tracking: {0}")]
+    EnableDirtyPageTracking(#[source] std::io::Error),
+    ///
+    /// Error disabling dirty page tracking
+    ///
+    #[error("Failed to disable dirty page tracking: {0}")]
+    DisableDirtyPageTracking(#[source] std::io::Error),
```
I believe these are obsoleted.

likebreath commented on vmm/src/lib.rs:
```diff
@@ -1064,7 +1064,6 @@ impl Vmm {
                     "Error during config migration"
                 )));
             }
-
```
Nit: Not needed.

likebreath commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -34,6 +34,7 @@ use std::fs::File;
 use std::os::unix::io::AsRawFd;
```
Please update the commit and make sure it is consistent with previous commits, e.g. "hypervisor: Implement start/stop_dirty_log for MSHV".

likebreath commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -16,7 +16,8 @@ pub use mshv_bindings::*;
 pub use mshv_ioctls::IoEventAddress;
```
Please update the commit message and make sure it consistent with previous commits: e.g., "hypervisor: Keep track of dirty memory slots for MSHV"

likebreath reviewed (commented):
@russell-islam I think the PR now is in a great shape. Other than a correction of the "readonly" check in `create_user_memory_region()` and a question from "stop_dirty_log", the rest are mostly cosmetic comments for cleanup and nit.  Any questions, please let me know.

With these being fixed, I will approve the PR and request one more review from Rob/Seb. I think we will land this PR very soon. Thanks for your hard work.

likebreath commented:
@russell-islam Can you please also update the PR description with some details about the experiments you have performed with this PR on MSHV live-migration? Thanks.

russell-islam commented:
@likebreath I have addressed all of your comments. It was a bumpy ride as we had to make lots of back and forth changes both for MSHV and KVM. Hope it was not too awful.

likebreath commented:
> It was a bumpy ride as we had to make lots of back and forth changes both for MSHV and KVM. Hope it was not too awful.

@russell-islam Agree. After all, I think we made the `hypervisor` crate interfaces cleaner (at least for live-migration support). Good learning for me to know a litter more on MSHV behaviors. Thanks again for your efforts and persistence throughout the process.

I just approved the PR :). Please review the commit message/contents one more time. Here are few typos I spotted:

commit 1 (b704557fb772f416f8239bd7f47ac88b67ecb834): obsoleted description "MSHV needs gpa and flags...Flags defines whether we want to clear or set the dirty bits in the hypervsior page state...";
commit 3 (7f25d9a2a498c9031b867f5bebb28b7374312657): need rewrite in general with typos, e.g. "disabledirty";
commit 4 (9cf8990430e0fda198fd608b31622bf22cbe5cd5): typo "secomp" (better to rename to "vmm: Add MSHV_VP_TRANSLATE_GVA to seccomp rule").

likebreath commented:
@liuw @rbradford @sboeuf I think the PR is ready for your final review. PTAL. Thanks.

sboeuf reviewed (approved):
Looks good!

liuw commented on hypervisor/src/mshv/mod.rs:
```diff
@@ -807,6 +814,17 @@ impl vm::Vm for MshvVm {
+        // No matter read only or not we keep track the slots.
+        // For readonly hypervisor can enable the dirty bits,
+        // but a VM exit happens before setting the dirty bits
+        self.dirty_log_slots.write().unwrap().insert(
```
Please remove this and the other `unwrap` in `remove_user_memory_region` in a follow-up PR.

rbradford replied:
Throughout the code we consider it acceptable to `.unwrap()` on a `MutexGuard` or `RwLockGuard`. panic()ing on there is the best cause of action since a user won't be able to do anything useful with an error from that.

liuw replied:
Fair enough. In that case, there is no need to change this.

liuw reviewed (approved):
Nice work and I'm super happy to see the collaboration.

I have one minor comment which can be addressed in follow-up PR.

liuw commented:
@russell-islam


I tried to resolve the conflict for you but GitHub's web UI created a merge commit by default, not fixing up the old commit. That stops this PR from being merged.

Please rebase and resolve that minor conflict. Please also take the chance to address my comment regarding `unwrap`.

russell-islam commented:
> @russell-islam
> 
> I tried to resolve the conflict for you but GitHub's web UI created a merge commit by default, not fixing up the old commit. That stops this PR from being merged.
> 
> Please rebase and resolve that minor conflict. ~Please also take the chance to address my comment regarding `unwrap`.~

Ignored your commit and rebased on my end.
